{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "ideal-precipitation",
   "metadata": {},
   "source": [
    "# Introduction"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "complex-wheat",
   "metadata": {},
   "source": [
    "## Package features"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "rubber-sponsorship",
   "metadata": {},
   "source": [
    "This package provides functionality to characterize sets of genomic regions. One or more sets of genomic regions are analyzed for the strength and significance of their association with certain biological features. This includes adjacency to genes from meaningful genesets and overlap with genomic regions such as ChIP-Seq peaks or predicted TFBS.\n",
    "\n",
    "Common to all analyses in this package is the notion of comparing a set of regions of interest (ROIs) with a set of background regions, in order to find out whether a biological feature is enriched or depleted in the ROIs compared to the background. Different scenarios are supported:\n",
    "\n",
    "1. Comparison within an experimental universe. In modern studies, we often have large datasets covering many considerably different biological populations (or single cells which can be grouped in various distinct clusters). For example, a dataset may characterize various distinct hematopoietic populations. Such studies typically aim to find and characterize various sets of ROIs characterizing different populations. In these cases, often the set of all identified ROIs can serve as a background against which individual ROI subsets are compared. Such tests answer the question: what distinguishes the ROI set characterizing population A from the ROIs observed in other populations?\n",
    "2. When the experimental design is not compatible with performing comparisons within the universe of experimentally observed ROIs, two approaches are supported:\n",
    "    1. Translate the observed ROIs into a set of associated genes and perform standard overrepresentation enrichment analysis.\n",
    "    2. Compare the observed ROIs against a simulated background (work in progress)."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "addressed-jason",
   "metadata": {},
   "source": [
    "## Package maturity"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "sublime-reaction",
   "metadata": {},
   "source": [
    "This package is unreleased and unpublished software, but we use it often in in-house projects. We can not yet provide support for external users. Also, we do change the API from time to time, without regard for health and safety of external users. Given this development stage, we also do not yet provide a complete documentation."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "legal-mongolia",
   "metadata": {},
   "source": [
    "## Content of this vignette"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "atmospheric-chancellor",
   "metadata": {},
   "source": [
    "This vignette is not meant as a complete documentation of all package features (cf. package maturity above). Rather, this notebook is intended to showcase how the package was used to perform some recently published analyses, in order to make the performed analyses more transparent."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "blind-payment",
   "metadata": {},
   "source": [
    "# Imports"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "following-raleigh",
   "metadata": {},
   "source": [
    "Run these imports prior to all examples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "modular-eight",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "\n",
    "import region_set_profiler as rsp\n",
    "\n",
    "import codaplot as co\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "from statsmodels.stats.multitest import multipletests"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "rotary-central",
   "metadata": {},
   "outputs": [],
   "source": [
    "%matplotlib inline"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "suspected-pathology",
   "metadata": {},
   "outputs": [],
   "source": [
    "def mht_correction(p_value_df):\n",
    "    pvals_corrected = multipletests(\n",
    "        p_value_df.to_numpy().flatten(order=\"C\"),\n",
    "        method=\"fdr_bh\",\n",
    "        returnsorted=False,\n",
    "    )[1]\n",
    "    q_value_df = pd.DataFrame(\n",
    "        pvals_corrected.reshape(p_value_df.shape),\n",
    "        index=p_value_df.index,\n",
    "        columns=p_value_df.columns,\n",
    "    )\n",
    "    return q_value_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "concerned-university",
   "metadata": {},
   "source": [
    "# Enrichments within an experimental universe"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "amino-zealand",
   "metadata": {},
   "source": [
    "See introduction above - these tools are meant to characterize subsets of large sets of ROIs against each other."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "expanded-minority",
   "metadata": {},
   "source": [
    "## Test for association with genesets"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "emotional-improvement",
   "metadata": {},
   "source": [
    "In this analysis type, we are interested in testing whether subsets of ROIs from a large experimental universe are significantly associated with certain genesets, i.e. whether the ROI subsets are in proximity of the genes of certain genesets more often than the background regions."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "horizontal-makeup",
   "metadata": {},
   "source": [
    "### Preparation of input data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "advised-serum",
   "metadata": {},
   "source": [
    "What is sufficient proximity to associate a genomic region with a certain gene? It will depend on the question at hand. Therefore, regionset_profiler makes no attempt at providing a general solution. Rather, the user provides a gene annotation for all ROIs, and decides about how to filter for biological meaningful proximity. regionset_profiler expects only the following information as a result of this analysis: a Series mapping each ROI to be used in the test to a gene."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "adverse-going",
   "metadata": {},
   "source": [
    "For this vignette, we provide 3,000 clustered toy DMRs (differentially methylated regions), placed (with considerable noise) in proximity of characteristic genes of four hematopoietic populations (real world datasets in this analysis type would be much larger, of course). Can we find out which DMR cluster belongs to which population?"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cutting-springer",
   "metadata": {},
   "source": [
    "For this example, gene annotation has already been performed. You can use any decent gene annotation tool of your liking and adapt gene annotations in any way you see fit to get a set of ROI-to-gene annotations with some confidence. In this case, we used our in-house tool [gftanno](https://github.com/stephenkraemer/gtfanno), but again this should be chosen on a case-by-case basis. "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "christian-aviation",
   "metadata": {},
   "source": [
    "Here are the annoteted DMRs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "unknown-italic",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>Gene</th>\n",
       "      <th>GenomicRegion</th>\n",
       "      <th>ClusterID</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"5\" valign=\"top\">1</th>\n",
       "      <th>7139742</th>\n",
       "      <th>7139990</th>\n",
       "      <td>Gm26983</td>\n",
       "      <td>Promoter</td>\n",
       "      <td>C1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9548071</th>\n",
       "      <th>9548115</th>\n",
       "      <td>Adhfe1</td>\n",
       "      <td>Promoter</td>\n",
       "      <td>C2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10316090</th>\n",
       "      <th>10316863</th>\n",
       "      <td></td>\n",
       "      <td>intergenic</td>\n",
       "      <td>C2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16297792</th>\n",
       "      <th>16297835</th>\n",
       "      <td>Stau2</td>\n",
       "      <td>intron</td>\n",
       "      <td>C3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33482222</th>\n",
       "      <th>33482405</th>\n",
       "      <td>Gm29228</td>\n",
       "      <td>Promoter</td>\n",
       "      <td>C2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                 Gene GenomicRegion ClusterID\n",
       "Chromosome Start    End                                      \n",
       "1          7139742  7139990   Gm26983      Promoter        C1\n",
       "           9548071  9548115    Adhfe1      Promoter        C2\n",
       "           10316090 10316863             intergenic        C2\n",
       "           16297792 16297835    Stau2        intron        C3\n",
       "           33482222 33482405  Gm29228      Promoter        C2"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "simulated_dmr_clusters = pd.read_pickle(\"simulated_dmr_clusters.p\")\n",
    "simulated_dmr_clusters.head(5)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cooked-rendering",
   "metadata": {},
   "source": [
    "This dataframe contains annotations to a gene (if possible) and genomic region (promoter, intron, exon etc.) as described here: [gtfanno](https://github.com/stephenkraemer/gtfanno). It's up to you to decide which annotations you want to consider. All of them? Or just the promoter annotations? Let's use everything within a gene or within a Promoter, and exclude everything else (for example intergenic DMRs which could not be associated with a gene). The further away from a gene we are, the less confidence we have in the annotations... Of course, if you had 4C data etc., you would integrate this at this step to get better annotations!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "further-trash",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>Gene</th>\n",
       "      <th>GenomicRegion</th>\n",
       "      <th>ClusterID</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"5\" valign=\"top\">1</th>\n",
       "      <th>7139742</th>\n",
       "      <th>7139990</th>\n",
       "      <td>Gm26983</td>\n",
       "      <td>Promoter</td>\n",
       "      <td>C1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9548071</th>\n",
       "      <th>9548115</th>\n",
       "      <td>Adhfe1</td>\n",
       "      <td>Promoter</td>\n",
       "      <td>C2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16297792</th>\n",
       "      <th>16297835</th>\n",
       "      <td>Stau2</td>\n",
       "      <td>intron</td>\n",
       "      <td>C3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33482222</th>\n",
       "      <th>33482405</th>\n",
       "      <td>Gm29228</td>\n",
       "      <td>Promoter</td>\n",
       "      <td>C2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33594864</th>\n",
       "      <th>33596335</th>\n",
       "      <td>Prim2</td>\n",
       "      <td>intron</td>\n",
       "      <td>C3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <th>...</th>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"5\" valign=\"top\">9</th>\n",
       "      <th>121008086</th>\n",
       "      <th>121008374</th>\n",
       "      <td>Ulk4</td>\n",
       "      <td>intron</td>\n",
       "      <td>C3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>121379312</th>\n",
       "      <th>121379380</th>\n",
       "      <td>Trak1</td>\n",
       "      <td>Promoter</td>\n",
       "      <td>C2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>121896742</th>\n",
       "      <th>121897004</th>\n",
       "      <td>Ackr2</td>\n",
       "      <td>Promoter</td>\n",
       "      <td>C3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>123078354</th>\n",
       "      <th>123078533</th>\n",
       "      <td>Zdhhc3</td>\n",
       "      <td>intron</td>\n",
       "      <td>C2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>123186193</th>\n",
       "      <th>123186646</th>\n",
       "      <td>Cdcp1</td>\n",
       "      <td>intron</td>\n",
       "      <td>C2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2241 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                   Gene GenomicRegion ClusterID\n",
       "Chromosome Start     End                                       \n",
       "1          7139742   7139990    Gm26983      Promoter        C1\n",
       "           9548071   9548115     Adhfe1      Promoter        C2\n",
       "           16297792  16297835     Stau2        intron        C3\n",
       "           33482222  33482405   Gm29228      Promoter        C2\n",
       "           33594864  33596335     Prim2        intron        C3\n",
       "...                                 ...           ...       ...\n",
       "9          121008086 121008374     Ulk4        intron        C3\n",
       "           121379312 121379380    Trak1      Promoter        C2\n",
       "           121896742 121897004    Ackr2      Promoter        C3\n",
       "           123078354 123078533   Zdhhc3        intron        C2\n",
       "           123186193 123186646    Cdcp1        intron        C2\n",
       "\n",
       "[2241 rows x 3 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dmrs_filtered = simulated_dmr_clusters.loc[\n",
    "    simulated_dmr_clusters[\"GenomicRegion\"].isin(\n",
    "        [\"Promoter\", \"5'-UTR\", \"3'-UTR\", \"intron\", \"exon\"]\n",
    "    )\n",
    "]\n",
    "dmrs_filtered"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "african-scoop",
   "metadata": {},
   "source": [
    "These are the regions we want to use for our enrichment analysis!"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "covered-vertical",
   "metadata": {},
   "source": [
    "Next, we need to define the genesets we are interested in. These files should be in the common GMT format (which is for example the format used by MSIGDB). For this demo, we have prepared some [marker genesets for hematopoietic populations](marker-genes-100.gmt)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "familiar-proposal",
   "metadata": {},
   "source": [
    "### Test and quantify association"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "driving-switzerland",
   "metadata": {},
   "source": [
    "We start by generating a GenesetOverlapStats object (if we were interested in comparing against other genomic intervals instead of regions, we would create a OverlapStats object instead)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "indie-threat",
   "metadata": {},
   "outputs": [],
   "source": [
    "overlap_stats = rsp.GenesetOverlapStats(\n",
    "    annotations=dmrs_filtered[\"Gene\"].str.upper(), genesets_fp=\"marker-genes-100.gmt\"\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "clinical-footage",
   "metadata": {},
   "source": [
    "We can use the compute method to identify overlaps between ROIs and the genesets:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "eligible-promise",
   "metadata": {},
   "outputs": [],
   "source": [
    "overlap_stats.compute()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "comfortable-bradford",
   "metadata": {},
   "source": [
    "The result is a coverage df, detailing for each ROI whether it is in proximity to a gene from a geneset:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "latin-disco",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>B-cells</th>\n",
       "      <th>CFU-E</th>\n",
       "      <th>HSC</th>\n",
       "      <th>Neutrophils</th>\n",
       "      <th>Monocytes</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"5\" valign=\"top\">1</th>\n",
       "      <th>7139742</th>\n",
       "      <th>7139990</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9548071</th>\n",
       "      <th>9548115</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16297792</th>\n",
       "      <th>16297835</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33482222</th>\n",
       "      <th>33482405</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33594864</th>\n",
       "      <th>33596335</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <th>...</th>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"5\" valign=\"top\">9</th>\n",
       "      <th>121008086</th>\n",
       "      <th>121008374</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>121379312</th>\n",
       "      <th>121379380</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>121896742</th>\n",
       "      <th>121897004</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>123078354</th>\n",
       "      <th>123078533</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>123186193</th>\n",
       "      <th>123186646</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2241 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                B-cells  CFU-E  HSC  Neutrophils  Monocytes\n",
       "Chromosome Start     End                                                   \n",
       "1          7139742   7139990          0      0    0            0          0\n",
       "           9548071   9548115          0      0    0            0          0\n",
       "           16297792  16297835         0      0    0            0          0\n",
       "           33482222  33482405         0      0    0            0          0\n",
       "           33594864  33596335         0      0    0            0          0\n",
       "...                                 ...    ...  ...          ...        ...\n",
       "9          121008086 121008374        0      0    0            0          0\n",
       "           121379312 121379380        0      0    0            0          0\n",
       "           121896742 121897004        0      0    0            0          0\n",
       "           123078354 123078533        0      0    0            0          0\n",
       "           123186193 123186646        0      0    0            0          0\n",
       "\n",
       "[2241 rows x 5 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "overlap_stats.coverage_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "confidential-crack",
   "metadata": {},
   "source": [
    "There are no hits in the head and tail - how many are there overall?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "superior-cover",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "B-cells        16\n",
       "CFU-E           7\n",
       "HSC            21\n",
       "Neutrophils     9\n",
       "Monocytes       9\n",
       "dtype: int64"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "overlap_stats.coverage_df.sum()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "running-wells",
   "metadata": {},
   "source": [
    "Next, we use the cluster information to aggregate this hits per cluster"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "different-latvia",
   "metadata": {},
   "outputs": [],
   "source": [
    "cluster_overlap_stats = overlap_stats.aggregate(dmrs_filtered[\"ClusterID\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "committed-qualification",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>dataset</th>\n",
       "      <th>B-cells</th>\n",
       "      <th>CFU-E</th>\n",
       "      <th>HSC</th>\n",
       "      <th>Neutrophils</th>\n",
       "      <th>Monocytes</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cluster_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>C1</th>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>17</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C2</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>8</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C3</th>\n",
       "      <td>11</td>\n",
       "      <td>3</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "dataset     B-cells  CFU-E  HSC  Neutrophils  Monocytes\n",
       "cluster_id                                             \n",
       "C1                4      4   17            1          1\n",
       "C2                1      0    0            8          5\n",
       "C3               11      3    4            0          3"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cluster_overlap_stats.hits"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "civilian-modern",
   "metadata": {},
   "source": [
    "We already have an intuition now - look like cluster C3 is associated with B cell marker genes, cluster C1 with HSC markers and cluster C2 with Neutrophil markers. But this needs to be tested. Let's test this, using the two-sided Fisher's exact test by default!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "adjusted-special",
   "metadata": {},
   "outputs": [],
   "source": [
    "cluster_overlap_stats.test_per_cluster_per_feature()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "israeli-eight",
   "metadata": {},
   "source": [
    "The p-values are here:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "fabulous-hollow",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>dataset</th>\n",
       "      <th>B-cells</th>\n",
       "      <th>CFU-E</th>\n",
       "      <th>HSC</th>\n",
       "      <th>Neutrophils</th>\n",
       "      <th>Monocytes</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cluster_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>C1</th>\n",
       "      <td>0.445164</td>\n",
       "      <td>0.251156</td>\n",
       "      <td>0.000022</td>\n",
       "      <td>0.172590</td>\n",
       "      <td>0.172590</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C2</th>\n",
       "      <td>0.028799</td>\n",
       "      <td>0.103667</td>\n",
       "      <td>0.000282</td>\n",
       "      <td>0.000826</td>\n",
       "      <td>0.163533</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C3</th>\n",
       "      <td>0.004754</td>\n",
       "      <td>0.686885</td>\n",
       "      <td>0.245615</td>\n",
       "      <td>0.065344</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "dataset      B-cells     CFU-E       HSC  Neutrophils  Monocytes\n",
       "cluster_id                                                      \n",
       "C1          0.445164  0.251156  0.000022     0.172590   0.172590\n",
       "C2          0.028799  0.103667  0.000282     0.000826   0.163533\n",
       "C3          0.004754  0.686885  0.245615     0.065344   1.000000"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cluster_overlap_stats.cluster_pvalues"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "suffering-azerbaijan",
   "metadata": {},
   "source": [
    "And here are the log-odds ratios:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "thorough-pioneer",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>dataset</th>\n",
       "      <th>B-cells</th>\n",
       "      <th>CFU-E</th>\n",
       "      <th>HSC</th>\n",
       "      <th>Neutrophils</th>\n",
       "      <th>Monocytes</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cluster_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>C1</th>\n",
       "      <td>-0.506878</td>\n",
       "      <td>1.202550</td>\n",
       "      <td>2.751653</td>\n",
       "      <td>-1.299779</td>\n",
       "      <td>-1.299779</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C2</th>\n",
       "      <td>-1.975638</td>\n",
       "      <td>-1.969892</td>\n",
       "      <td>-3.442843</td>\n",
       "      <td>3.221577</td>\n",
       "      <td>1.305871</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C3</th>\n",
       "      <td>2.104289</td>\n",
       "      <td>0.767077</td>\n",
       "      <td>-0.769356</td>\n",
       "      <td>-2.243714</td>\n",
       "      <td>0.279752</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "dataset      B-cells     CFU-E       HSC  Neutrophils  Monocytes\n",
       "cluster_id                                                      \n",
       "C1         -0.506878  1.202550  2.751653    -1.299779  -1.299779\n",
       "C2         -1.975638 -1.969892 -3.442843     3.221577   1.305871\n",
       "C3          2.104289  0.767077 -0.769356    -2.243714   0.279752"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cluster_overlap_stats.log_odds_ratio"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "typical-winner",
   "metadata": {},
   "source": [
    "Finally, perform MHT correction:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "collaborative-brooks",
   "metadata": {},
   "outputs": [],
   "source": [
    "q_values = mht_correction(cluster_overlap_stats.cluster_pvalues)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "republican-plane",
   "metadata": {},
   "source": [
    "### Visualize the results"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "affecting-rates",
   "metadata": {},
   "source": [
    "#### Log-odds"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "electric-serial",
   "metadata": {},
   "source": [
    "Let's first look at the log-odds ratios:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "desperate-updating",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAVgAAAEoCAYAAAAdTumBAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjMuNCwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy8QVMy6AAAACXBIWXMAAAsTAAALEwEAmpwYAAAXi0lEQVR4nO3deZRcZZ3G8efpTgeQgCwCUcABkUUNiWxxUBgEZVHPqCgqCgKK4jYaD4yDDDNOkFH0MGwiinIAAwoMCtEgUVCDg7IpYU1gkEVGVpkRIwhi0t2/+ePeTopQ6a5U3bfuwvdzTp1TdW9X5Wmlf/3r977vex0RAgAUb6DsAADQVBRYAEiEAgsAiVBgASARCiwAJEKBBYBEKLAAkLO9pu1f2b7V9mLbx/X0ecyDBYCMbUtaOyL+bHtI0i8lzYqI67v5vEmFpgOAGous4/xz/nIof3TdhSYvsNfstnvjWuTXnPbZsiMkcdayV5UdoXCzPvb5siMk8dSVs8uOkMTQRi91r5/xUW/RUc05M+5v+2/ZHpS0UNLLJZ0RETd0m4UxWABoEREjEfFqSZtJmml7WrefRYEF0CiD7uwxkYhYIukqSft1m4UCC6BRBu2OHu3Y3sj2evnztSTtLem/u83CRS4AjdJJdzqOF0uak4/DDki6OCJ+2O2HUWABNMqqutNORMRtknYoKgsFFkCj9NjBFooCC6BReulgi0aBBdAodLAAkAgdLAAkUqW5pxRYAI1CBwsAiTAGCwCJVKmDrdJwBQA0Ch0sgEaZPFCdDpYCC6BRGIMFgESqNAZLgQXQKFXqYLnIBQCJ0MECaJQqdbAUWACNwiwCAEiEi1wAkEiVhgi4yAUAidDBAmgUhggAIBEucgFAIgN0sACQhit0lYuLXACQCB0sgEYZqFAHS4EF0CgerM4f5h0VWNtTJZ0qaRdJSyT9XtKXJZ2ibJhhSNLpEXFmkpQA0KHByTUqsLYtaa6kORFxYH5shqT1JO0aEX+1PUXSItvzIuLhlIEBYDweqFGBlbSnpGWt3WlE3LrS16whLpgBwLN0UmCnSVrY7oTtzSVdLunlkj5D9wqgbFW6yNVT1xkRD0TEdGUF9lDbmxQTCwC640F39OiHTgrsYkk7jfcFeee6SNLuRYQCgG55cKCjRz908q8skLSG7SPGDtiebnt322vlr9eXtJuku9LEBIDODA4NdPTohwnHYCMibO8v6VTbR0t6RtL9kr4v6QzbIcmS/iMibk+YFQBqpaN5sPkQwLvbnDqr2DgA0JuBCi00qE4SAChALxe5bG9u+yrbd9hebHtWL1lYKgugUXqcITAs6aiIuMn2OpIW2v5JRNzRzYdRYAE0yuDkwa7fGxGPSHokf/6k7TslbSqpqwLLEAGARtn04EM188cLtOnBh457XtLs8T7H9haSdpB0Q7dZ6GABNMojF5ynRy44T5I00Ob2MWPnd57/s9mr+ox8f5VLJH06Ip7oNgsFFkCj9LqIwPaQsuL6nYi4tJfPosACaJRe9iLIdw88W9KdEXFyz1l6/QAAqJIe9yJ4naT3S9rL9i35483dZqGDBdAovQwRRMQvla1MLQQFFkCjVGm7QgosgEZxm5kDZaHAAmiUKu1FQIEF0Cj92ky7ExRYAI1Su9t2A0Bd1O2usgBQG4zBAkAiDBEAQCIUWABIhDFYAEjEg91vuF00CiyARmGIAAASGWCIAADSoIMFgESqVGCrkwQAGoYOFkCjDAxVp6xVJwkAFKBKQwQUWACNwl4EAJBIlVZyVScJADQMHSyARhmYXJ2yVp0kAFCAKg0RJC+we/1lu9T/RP8d8f2yEyQRo5eUHaFwH7xlQdkRknhq7tSyIySx3hFf7PkzBtjsBQDSqNI0reokAYCGoYMF0ChV6mApsAAahaWyAJAIHSwAJFKlaVrVSQIADUMHC6BRPMA8WABIgwILAGl4aKjsCMsxBgugWQYGO3u0Yfsc24/ZXlRIlCI+BAAa4luS9ivqwxgiANAsPYzBRsTVtrcoKgoFFkCjDG07U5O23kXDd/9aw3ff+Jzzk7beWZO23kWSZuePZCiwABpl+Le3afi3t2UvJk1e5fk19zl8duosFFgAjcJKLgB4HqDAAmiW3qZpXSjpOknb2n7Q9uG9RGGIAECz9DaL4L0FJqHAAmgWc08uAEhkEktlAaDx6GABNArbFQJAKhWaB0uBBdAodLAAkAoFFgASYYgAANJgHiwApMIQAQAkQoEFgDRqt12h7am2L7J9r+2Ftufbnmn7OtuLbd9m+z2pwwLAhHrYTatoE3awti1prqQ5EXFgfmyGpPUkHRIRd9t+iaSFtq+IiCUJ8wLA+FydDraTIYI9JS2LiDPHDkTEra1fEBEP235M0kaSlhSaEABWR80K7DRJC8f7AtszJU2WdG8RoQCgW1GzAjsu2y+WdL6kQyNitPdIANCDChXYTpIslrRTuxO215V0uaRjI+L6IoMBQFfszh590EmBXSBpDdtHjB2wPd32Hsoufp0XEd9LFRAAVsvAQGePPphwiCAiwvb+kk61fbSkZyTdL+l6SX8naUPbh+VfflhE3JImKgBMrHZjsBHxsKR3tzl1fLFxAKBHdSuwAFAbFFgASIQCCwBp1G4MFgBqgwILAIn0aY5rJyiwAJqFDhYA0qjSGGx1kgBAw9DBAmiWweqUteokAYAiVGiIgAILoFkosACQBhe5AOB5gA4WQLPQwQJAIj3ettv2frbvsn2P7c/2EoUOFkCj9DIGa3tQ0hmS9pb0oKRf254XEXd083l0sACaxQOdPdqbKemeiLgvIpZKukjS27qNQoEF0CiTJk3SmmutpUmT2v+BPnZe0uw2pzeV9EDL6wfzY91l6faNAFBFS4dHtHR4JH/13J21xs6/YK01Z6fOQoEF0CgjEb28/SFJm7e83iw/1hWGCAA0ymh09liFX0va2vaWtidLOlDSvG6z0MECaJSRcarnRCJi2PY/SLpC0qCkcyJicbefR4EF0Cg9DRBIioj5kuYXkYUCC6BRemhgC0eBBdAo0dtFrkJRYAE0yrLRshOsQIEF0Ch0sACQSIUaWAosgGbpZZpW0SiwABqlOuWVAgugYSrUwFJgATTLcIUqbPICe+XFJ6T+J/pup+u/VnaEJE778PllRyjc/BPOKTtCEm95vOwEaVxTwGdUqL7SwQJolgrN0qLAAmiW0Qpd5qLAAmgUpmkBQCLLRiiwAJBEj3c0KBQFFkCj0MECQCKjdLAAkAZDBACQSIUmEVBgATTLM8PV2bCQAgugURiDBYBEWGgAAIkso8ACQBoMEQBAIqN0sACQxjMjzCIAgCToYAEgkQptRUCBBdAsXOQCgESWMgYLAGksZaksAKTBSi4ASIQOFgASqVIHO1B2AAAo0shodPRYXbbfZXux7VHbO3fyHjpYAI2SsINdJOkdkr7R6RsosAAa5emlI0k+NyLulCTbHb+HIQIAjbLvthvrtP23177bbtx2aGDsvKTZqbPQwQJolMvueFSX3fHohOe/fsCM2Sufs/1TSVPbvO3YiPjB6mahwAJolF6maUXEGwuM0tkQge2pti+yfa/thbbn297G9o9tL7H9wyJDAUC3Us0i6MaEBdbZiO5cST+PiK0iYidJx0jaRNKJkt6fNiIAdG5kdLSjx+qyvb/tByXtKuly21dM9J5Ohgj2lLQsIs4cOxARt7b8o69f7aQAkEjCWQRzlTWbHeukwE6TtLCrRADQZ1VaycVFLgCNUrcCu1jSAamDAEARqlRgO5lFsEDSGraPGDtge7rt3dPFAoDuLB0e7ejRDxMW2IgISftLemM+TWuxpBMkPWr7F5K+K+kNth+0vW/auAAwvpHh0Y4e/dDRGGxEPCzp3W1O0cUCqJTgnlwAkEa/utNOUGABNEpUp75SYAE0C0MEAJDIaIWmaVFgATTK8LI0S2W7QYEF0ChBBwsAaVBgASCRkZHqTCOgwAJoFKZpAUAiLDQAgEQYIgCARLjIBQCJUGABIBFWcgFAIqOMwQJAGiPDw2VHWI4CC6BRYpS9CAAgidFlS8uOsBwFFkCj0MECQCIUWABIhAILAIkML/1L2RGWo8ACaBQ6WABIhAILAIkwTQsAEqGDBYBERoeXlR1hOQosgEYZHWaIAACSGGWIAADSiJE0Bdb2iZL+XtJSSfdK+kBELBnvPQNJkgBASWJ0pKNHF34iaVpETJf0G0nHTPQGOlgAjZJqDDYirmx5eb2kAyZ6jyOqc3sFAKgD25dJ+s+I+PZ4X0cHCwA52z+VNLXNqWMj4gf51xwraVjSdyb8PDpYAOiM7cMkfUTSGyLi6Ym+ng4WADpgez9J/yRpj06Kq0QHCwAdsX2PpDUk/SE/dH1EfHTc91BgASCN2s+Dtf0622vnzw+2fbLtvyk7Vzdsb9fyfI2Vzv1t/xMVw/aatjdqc3wj22uWkQmdsz1ge92yc9RR7QuspK9Letr2DElHKVthcV65kbp2Qcvz61Y697V+BinYVyTt3ub4bpJO6XOWQtmeZXtdZ862fZPtfcrO1SvbF+Tf19qSFkm6w/Znys5VN00osMORjXO8TdJXI+IMSeuUnKlbXsXzdq/rZKeIuHTlgxExV9LflZCnSB+MiCck7SNpfUnvl/SlciMV4pX59/V2ST+StKWy7w2roQkF9knbx0g6WNLltgckDZWcqVuxiuftXtfJC8Y5V/f/Bsd+8b1Z0vkRsVj1/mU4Zsj2kLICOy8ilqne/w2WognTtN4j6X2SDo+IR22/VNKJJWfq1ma2v6LsB3TsufLXm5YXq2eP2Z4ZEb9qPWh7F0n/W1Kmoiy0faWyDu8Y2+tIGi05UxG+Iel+SbdKujq/rvFEqYlqiFkEFWL70PHOR8ScfmUpku2Zki6W9C1JC/PDO0s6RNKBEXFDSdF6lv/F9GpJ90XEEtsbSto0Im4rN1nxbE+KiOGyc9RJbQus7SfV/k8WS4qIaMRVT9tTI+LRsnP0yvbGkj4haVp+aJGkMyLisfJSdc/2juOdj4ib+pUlBdubSPqipJdExJtsv1LSrhFxdsnRaqW2Bfb5wvZNETHuD3Pd2J4s6VWSHqpxgb1qnNMREXv1LUwCtn8k6Vxla/Bn2J4k6eaI2L7kaLVS2zFY2xuMdz4iHu9XlsRqf8HE9pmSTo+IxbZfqGwK2oikDWz/Y0RcWG7C1RcRe5adIbEXRcTF+QVkRcSw7ercKqAmaltglY3lhdoXoJD0sv7GSeassgMUYPeWJYUfkPSbiHi77anKpgDVrsDa3isiFth+R7vz7aal1cxT+XhySMsXuvyp3Ej1U9sCGxFblp2haG268lC2kKLuWndA3lvSdyUpn/VRTqLe7SFpgbJbiKwsJNW9wB4paZ6krWxfI2kjSe8qN1L91H4M1tlP6EGStoyI4/NpWlNXnhJUB7Z/q+d25VOUTZX5UETcX0auXuXjlSdJekjSVZK2y4vrJEmLImK7cT8AfZcv1R6RtK2y/x7vkjQQEX8tNVjNNKHAfl3ZvMO9IuIVtteXdGVE7FJytMLkf4YeERH7lZ2lG7a3UbZcdqqkUyPiW/nxfSXtExFHlRivJ3kheqekLdTyF2FEfL6sTEVod3G1iRdcU6vtEEGL10TEjrZvlqSI+GN+lboxIuJS2/9Sdo5uRcRvJD3nl0NEXCHpiv4nKtQPlI1NLpRU++4uHxffVNJatnfQir+m1tX4K/LQRhMK7DLbg1oxGL+RmrGSZjnbU1TjJaW2T9c4yywj4lN9jFO0zer6l8Uq7CvpMEmbSTq55fgTkv65jEB11oQC+xVJcyVtbPsLyu70WMtuz/aRbQ6vL+mtkr7a5zhFurHl+XGS/q2sIAlca3v7iLi97CBFyFcLzrH9zoi4pOw8dVf7MVhp+T6qb1D258zPIuLOkiN1xfbKhSeU7Z5+dVN+gG3fHBE7lJ2jV7ZvV/b/zyRJW0u6T9kQwdhKwuklxuuZ7fskXSLpnLr+PFVB7QtsPj9vcUQ8mb9eV9Ir6ri+/fmw1rspF0om2tQ9Iv6nX1lSyDetOVDZvOUBSedIuijfwhAdakKBvVnSjvmesGObb9xYxx/i1uJj+/SI+GTZmYrWlALbKt+XYDdlHe01dd+HYGW291C2Gfx6kr4n6fiIuKfUUDVR2wsnLRwtvyUiYlT1HVtunf/6utJSFMz2k7afsP2EpOljz8eOl52vF7Y/J2mOpA0lvUjSuXWe8THG9qDtt9qeK+lUZfOYXybpMknzy8xWJ3UtRK3us/0prVjx9HFl42F1VO8/J1YhIup6h4lOHCRpRkQ8I0m2vyTpFkn/XmaoAtytbFHIiRFxbcvx79mu+10o+qYJQwQbK5tJsJeyAvUzSZ+u4y5Ntp+WdI+yTnar/Plydb9w0kT5KrX9I2JJ/no9SZc2YDetKRHx57Jz1F3tC2yT2N5a0iaSHljp1OaSHmXcq3psf1/SLpJ+ouwX/N6SfiXpQam+c3xtz5E0q+UXx/qSToqID5YarGaaMESwXAMuoJwi6ZiVr0DnMyNOUfuNRVCuufljzM9LylG06WPFVVq+QrL20+v6rVEFVvXfO3WTdvNdI+J221uUkAcTiIg5+dLsbfJDd+U3CKy7AdvrR8QfpeU7vTWtXiTXtP/BLi87QI/WG+fcWv0Kgc7Zfr2yWQT3K/sFv7ntQyPi6hJjFeEkSdfZ/q6y7+sASV8oN1L9NGoM1vaLJP0havpN2b5Q0oKIOGul4x+StHdEvKecZFgV2wslvS8i7spfbyPpwojYqdxkvcvvwzV28fiqiLij5Ei1U9sONl/B9SVJj0s6XtL5yuYhDtg+JCJ+XGa+Ln1a0lzbB+nZd1+dLGn/skJhXENjxVXKdg6zPVRmoAINacWwW1O+p76qbQdr+0Zlu/u8UNI3Jb0pIq7P9yW4sM7r3W3vqRV3X10cEQvKzINVs32uso2pv50fOkjSYN2vttueJenDyvYjsLJf8N+MiNNLDVYzdS6wt0TEq/Pnd0bEK1rONWJDEVRfvuH2J5QtlZWkX0j6Wt13/rd9m7LbdD+Vv15b0nXMxV49tR0i0LP3fP3LSufq+VsDtZLvQ3xrfsubkyf6+pqxss58zIjqP0un7+pcYGfk69itbPf1sTXtlrRmebHwfBERI7bvsv3SiPhd2XkKdq6kG/K9CCTp7ZLOLi9OPdV2iACoAttXS9pB2eqtp8aOR8RbSwtVENs7acWmQ7+IiJvLzFNHFFigB/lWfs8REf/V7yxFy4dANtGzb+bYtE49qToPEQBV8OaIOLr1gO0vS6p1gbX9SWW39vm9Voy/hiQucq0GOligB6u4vfVtdb/abvseZXds/kPZWeqMDhbogu2PKdt7eKt8StOYdSRd2/5dtfKAstuRowd0sEAXbL9Q2R1/T5D02ZZTT0bE4+WkKo7tsyVtq2x/j+VzeiOiadPRkqKDBboQEX+S9CfbR690akq+WXXdLwb9Ln9Mzh/oAh0s0IOW23ePzb/eUtmWha8qNVhBbE+RJO5u0B06WKAHEbF96+v8DrMfLylOYWxPU7aB0gb56/+TdEhELC41WM3QwQIFs337yoW3bmxfK+nYiLgqf/16SV+MiNeWmatu6GCBHtg+suXlgKQdJT1cUpwirT1WXCUpIn6eb/iC1UCBBXrTekvyYWVX3S8pKUuR7rP9r8qGCSTpYEn3lZinlhgiAApg+wUR8XTZOYqS30X2OD17G8bZY/foQmcosEAPbO+qbJepKRHxUtszJH0kImp/oQu9o8ACPbB9g7IbAs4b2+Td9qKImDb+O6vJ9rzxzjdhl7B+YgwW6FFEPGA/ay/qkVV9bQ3sqmyZ7IWSbhCbbPeEAgv05gHbr5UU+c0OZ0m6s+RMvZgqaW9J75X0PmUX7S5k/mt3GCIAepDfKv40SW9U1u1dKWlWE3ahyu839l5JJ0o6LiK+WnKk2qHAAniWvLC+RVlx3ULSPEnnRMRDZeaqIwos0AXbnxvndETE8X0LUyDb5ym7Zfx8SRdFxKKSI9UaBRbogu2j2hxeW9LhkjaMiCl9jlQI26NacW+x1uJgZb841u1/qvqiwAI9sr2Osotbh0u6WNJJEfFYualQBcwiALpkewNJR0o6SNIcSTuy0gmtKLBAF2yfKOkdkr4paXv2S0U7DBEAXcjHKv+qbIMXxirRFgUWABIZKDsAADQVBRYAEqHAAkAiFFgASOT/AYnetbsHjwLCAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<Figure size 432x288 with 2 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "fig, ax = plt.subplots(1, 1)\n",
    "_ = co.heatmap(\n",
    "    df=cluster_overlap_stats.log_odds_ratio,\n",
    "    ax=ax,\n",
    "    xticklabels=True,\n",
    "    yticklabels=True,\n",
    "    yticklabel_side='left',\n",
    "    cmap=\"RdBu_r\",\n",
    "    show_guide=True,\n",
    "    norm=co.plotting.MidpointNormalize(vmin=-2, midpoint=0, vmax=3),\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "abandoned-chart",
   "metadata": {},
   "source": [
    "#### P-values"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "charged-basin",
   "metadata": {},
   "source": [
    "To confirm the trends we see in the log-odds plot, let's visualize the p-values. One helpful visualization is a heatmap of `-np.sign(cluster_overlap_stats.log_odds_ratio) * np.log10(q_values)`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "gorgeous-ordinary",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAVgAAAEoCAYAAAAdTumBAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjMuNCwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy8QVMy6AAAACXBIWXMAAAsTAAALEwEAmpwYAAAXi0lEQVR4nO3deZhcVZ3G8fftTgKRRVCBKOCACLgAYZGMqIigLDrjAm4oCCgOOjqKj46jiKNhcH2QRRFFHcDgAoNKFAQFNTioLGpkS2BQQEYQEUeNoIhJd//mj3s7KUKlu1J1T917D9/P89TzVN3bVXlb6V//+txzznVECABQvZG6AwBAriiwAJAIBRYAEqHAAkAiFFgASIQCCwCJUGABoGR7Xds/tn2d7aW2jxvo85gHCwAF25a0XkT82fZMST+UdHREXNXP582oNB0AtFgUHeefy5czy0ffXWjyAnvzUQdl1yJv9bH/rDtCElf9bqLuCJW74Z776o6QxOt3nlN3hCTWnT3bg37GG71VTzXn9Li9679le1TSYklPlHRaRFzdbxbGYAGgQ0SMR8TOkraQNM/2Dv1+FgUWQFZG3dtjOhGxTNJlkg7oNwsFFkBWRu2eHt3Y3sT2RuXz2ZL2lfQ//WbhIheArPTSnU7hsZIWlOOwI5LOi4hv9vthFFgAWVlTd9qLiLhe0i5VZaHAAsjKgB1spSiwALIySAdbNQosgKzQwQJAInSwAJBIk+aeUmABZIUOFgASYQwWABJpUgfbpOEKAMgKHSyArMwaaU4HS4EFkBXGYAEgkSaNwVJgAWSlSR0sF7kAIBE6WABZaVIHS4EFkBVmEQBAIlzkAoBEmjREwEUuAEiEDhZAVhgiAIBEuMgFAImM0MECQBpu0FUuLnIBQCJ0sACyMtKgDpYCCyArHm3OH+Y9FVjbcySdIml3Scsk/VbSRyWdrGKYYaakUyPi9CQpAaBHo7NaVGBtW9JCSQsi4uDy2FxJG0naIyL+Znt9SUtsXxARd6UMDABT8UiLCqykvSWt6OxOI+K61b5mHXHBDAAepJcCu4Okxd1O2N5S0kWSnijpnXSvAOrWpItcA3WdEXFHROykosAebnuzamIBQH886p4ew9BLgV0qabepvqDsXJdI2rOKUADQL4+O9PQYhl7+lUWS1rF91OQB2zvZ3tP27PL1xpKeJenmNDEBoDejM0d6egzDtGOwERG2D5R0iu13SXpA0u2Svi7pNNshyZI+FhE3JMwKAK3S0zzYcgjgFV1Ofa7aOAAwmJEGLTRoThIAqMAgF7lsb2n7Mts32l5q++hBsrBUFkBWBpwhMCbpHRHxM9sbSFps+zsRcWM/H0aBBZCV0Vmjfb83In4j6Tfl8/ts3yRpc0l9FViGCABkZfNDD9e8by/S5ocePuV5SfOn+hzbW0naRdLV/WahgwWQld98+Wz95stnS5JGutw+ZvL80y7+3vw1fUa5v8rXJL0tIu7tNwsFFkBWBl1EYHumiuL6pYg4f5DPosACyMogexGUuweeIemmiDhp4CyDfgAANMmAexE8U9JrJO1j+9ry8YJ+s9DBAsjKIEMEEfFDFStTK0GBBZCVJm1XSIEFkBV3mTlQFwosgKw0aS8CCiyArAxrM+1eUGABZKV1t+0GgLZo211lAaA1GIMFgEQYIgCARCiwAJAIY7AAkIhH+99wu2oUWABZYYgAABIZYYgAANKggwWARJpUYJuTBAAyQwcLICsjM5tT1pqTBAAq0KQhAgosgKywFwEAJNKklVzNSQIAmaGDBZCVkVnNKWvNSQIAFWjSEEHyAvuBZ78n9T8xdC+8/YG6IyRx6OHvrTtC5c7/7U/qjpDErMs/V3eENJ749IE/YoTNXgAgjSZN02pOEgDIDB0sgKw0qYOlwALICktlASAROlgASKRJ07SakwQAMkMHCyArHmEeLACkQYEFgDQ8c2bdEVZiDBZAXkZGe3t0YftM2/fYXlJJlCo+BAAy8XlJB1T1YQwRAMjLAGOwEXG57a2qikKBBZCVmdvP04xtd9fYL36isV/89CHnZ2z7NM3YdndJml8+kqHAAsjK2C+v19gvry9ezJi1xvPr7nfk/NRZKLAAssJKLgB4GKDAAsjLYNO0zpF0paTtbd9p+8hBojBEACAvg80ieFWFSSiwAPJi7skFAInMYKksAGSPDhZAVtiuEABSadA8WAosgKzQwQJAKhRYAEiEIQIASIN5sACQCkMEAJAIBRYA0mjddoW259g+1/atthfbvtj2PNtX2l5q+3rbr0wdFgCmNcBuWlWbtoO1bUkLJS2IiIPLY3MlbSTpsIj4he3HSVps+5KIWJYwLwBMzc3pYHsZIthb0oqIOH3yQERc1/kFEXGX7XskbSJpWaUJAWBttKzA7iBp8VRfYHuepFmSbq0iFAD0K1pWYKdk+7GSviDp8IiYGDwSAAygQQW2lyRLJe3W7YTtDSVdJOnYiLiqymAA0Be7t8cQ9FJgF0lax/ZRkwds72R7LxUXv86OiK+mCggAa2VkpLfHEEw7RBARYftASafYfpekByTdLukqSc+W9GjbR5RffkREXJsmKgBMr3VjsBFxl6RXdDl1fLVxAGBAbSuwANAaFFgASIQCCwBptG4MFgBagwILAIkMaY5rLyiwAPJCBwsAaTRpDLY5SQAgM3SwAPIy2pyy1pwkAFCFBg0RUGAB5IUCCwBpcJELAB4G6GAB5IUOFgASGfC23bYPsH2z7Vtsv3uQKHSwALIyyBis7VFJp0naV9Kdkn5i+4KIuLGfz6ODBZAXj/T26G6epFsi4raIWC7pXEkv7jcKBRZAVmbMmKF1Z8/WjBnd/0CfPC9pfpfTm0u6o+P1neWx/rL0+0YAaKLlY+NaPjZevnrozlqT5x8xe935qbNQYAFkZTxikLf/WtKWHa+3KI/1hSECAFmZiN4ea/ATSdva3tr2LEkHS7qg3yx0sACyMj5F9ZxORIzZ/hdJl0galXRmRCzt9/MosACyMtAAgaSIuFjSxVVkocACyMoADWzlKLAAshKDXeSqFAUWQFZWTNSdYBUKLICs0MECQCINamApsADyMsg0rapRYAFkpTnllQILIDMNamApsADyMtagCpu8wL5//+1T/xNDt/Vfbq07QhKLrl1Ud4Tq3Xh13QmSuHhF3QnS+McKPqNB9ZUOFkBeGjRLiwILIC8TDbrMRYEFkBWmaQFAIivGKbAAkMSAdzSoFAUWQFboYAEgkQk6WABIgyECAEikQZMIKLAA8vLAWHM2LKTAAsgKY7AAkAgLDQAgkRUUWABIgyECAEhkgg4WANJ4YJxZBACQBB0sACTSoK0IKLAA8sJFLgBIZDljsACQxnKWygJAGqzkAoBE6GABIJEmdbAjdQcAgCqNT0RPj7Vl++W2l9qesP20Xt5DBwsgKwk72CWSDpL0mV7fQIEFkJX7l48n+dyIuEmSbPf8HoYIAGRl/+031ccP3FH7b79p16GByfOS5qfOQgcLICsX3ni3Lrzx7mnPf/plc+evfs72dyXN6fK2YyPiG2ubhQILICuDTNOKiOdVGKW3IQLbc2yfa/tW24ttX2x7O9vftr3M9jerDAUA/Uo1i6Af0xZYFyO6CyV9PyK2iYjdJB0jaTNJJ0h6TdqIANC78YmJnh5ry/aBtu+UtIeki2xfMt17ehki2FvSiog4ffJARFzX8Y8+Z62TAkAiCWcRLFTRbPaslwK7g6TFfSUCgCFr0kouLnIByErbCuxSSS9LHQQAqtCkAtvLLIJFktaxfdTkAds72d4zXSwA6M/ysYmeHsMwbYGNiJB0oKTnldO0lkr6sKS7bf9A0lckPdf2nbb3TxsXAKY2PjbR02MYehqDjYi7JL2iyym6WACNEtyTCwDSGFZ32gsKLICsRHPqKwUWQF4YIgCARCYaNE2LAgsgK2Mr0iyV7QcFFkBWgg4WANKgwAJAIuPjzZlGQIEFkBWmaQFAIiw0AIBEGCIAgES4yAUAiVBgASARVnIBQCITjMECQBrjY2N1R1iJAgsgKzHBXgQAkMTEiuV1R1iJAgsgK3SwAJAIBRYAEqHAAkAiY8v/WneElSiwALJCBwsAiVBgASARpmkBQCJ0sACQyMTYirojrESBBZCViTGGCAAgiQmGCAAgjRhPU2BtnyDphZKWS7pV0msjYtlU7xlJkgQAahIT4z09+vAdSTtExE6Sfi7pmOneQAcLICupxmAj4tKOl1dJetl073FEc26vAABtYPtCSf8VEV+c6uvoYAGgZPu7kuZ0OXVsRHyj/JpjJY1J+tK0n0cHCwC9sX2EpDdIem5E3D/d19PBAkAPbB8g6d8k7dVLcZXoYAGgJ7ZvkbSOpN+Xh66KiDdO+R4KLACk0fp5sLafaXu98vmhtk+y/Xd15+qH7Sd1PF9ntXNPH36iathe1/YmXY5vYnvdOjKhd7ZHbG9Yd442an2BlfRpSffbnivpHSpWWJxdb6S+fbnj+ZWrnfvUMINU7BOS9uxy/FmSTh5ylkrZPtr2hi6cYftntverO9egbH+5/L7Wk7RE0o2231l3rrbJocCORTHO8WJJn4yI0yRtUHOmfnkNz7u9bpPdIuL81Q9GxEJJz64hT5VeFxH3StpP0saSXiPpI/VGqsRTyu/rJZK+JWlrFd8b1kIOBfY+28dIOlTSRbZHJM2sOVO/Yg3Pu71uk0dMca7t/w1O/uJ7gaQvRMRStfuX4aSZtmeqKLAXRMQKtfu/wVrkME3rlZJeLenIiLjb9uMlnVBzpn5tYfsTKn5AJ5+rfL15fbEGdo/teRHx486DtneX9LuaMlVlse1LVXR4x9jeQNJEzZmq8BlJt0u6TtLl5XWNe2tN1ELMImgQ24dPdT4iFgwrS5Vsz5N0nqTPS1pcHn6apMMkHRwRV9cUbWDlX0w7S7otIpbZfrSkzSPi+nqTVc/2jIgYqztHm7S2wNq+T93/ZLGkiIgsrnranhMRd9edY1C2N5X0Zkk7lIeWSDotIu6pL1X/bO861fmI+NmwsqRgezNJH5L0uIh4vu2nSNojIs6oOVqrtLbAPlzY/llETPnD3Da2Z0l6qqRft7jAXjbF6YiIfYYWJgHb35J0loo1+HNtz5B0TUTsWHO0VmntGKztR011PiL+MKwsibX+gont0yWdGhFLbT9SxRS0cUmPsv2vEXFOvQnXXkTsXXeGxB4TEeeVF5AVEWO2m3OrgJZobYFVMZYX6l6AQtIThhsnmc/VHaACe3YsKXytpJ9HxEtsz1ExBah1Bdb2PhGxyPZB3c53m5bWMn8px5NDWrnQ5U/1Rmqf1hbYiNi67gxV69KVh4qFFG3XuQPyvpK+IknlrI96Eg1uL0mLVNxCZHUhqe0F9u2SLpC0je0fSdpE0svrjdQ+rR+DdfETeoikrSPi+HKa1pzVpwS1ge1f6qFd+foqpsq8PiJuryPXoMrxyhMl/VrSZZKeVBbXGZKWRMSTpvwADF25VHtc0vYq/nu8WdJIRPyt1mAtk0OB/bSKeYf7RMSTbW8s6dKI2L3maJUp/ww9KiIOqDtLP2xvp2K57BxJp0TE58vj+0vaLyLeUWO8gZSF6KWStlLHX4QR8R91ZapCt4urOV5wTa21QwQd/j4idrV9jSRFxB/Lq9TZiIjzbb+37hz9ioifS3rIL4eIuETSJcNPVKlvqBibXCyp9d1dOS6+uaTZtnfRqr+mNtTUK/LQRQ4FdoXtUa0ajN9EeaykWcn2+mrxklLbp2qKZZYR8dYhxqnaFm39y2IN9pd0hKQtJJ3UcfxeSe+pI1Cb5VBgPyFpoaRNbX9QxZ0eW9nt2X57l8MbS3qRpE8OOU6Vftrx/DhJ768rSAJX2N4xIm6oO0gVytWCC2y/NCK+Vneetmv9GKy0ch/V56r4c+Z7EXFTzZH6Ynv1whMqdk+/PJcfYNvXRMQudecYlO0bVPz/M0PStpJuUzFEMLmScKca4w3M9m2SvibpzLb+PDVB6wtsOT9vaUTcV77eUNKT27i+/eGw1juXCyXTbeoeEf87rCwplJvWHKxi3vKIpDMlnVtuYYge5VBgr5G0a7kn7OTmGz9t4w9xZ/GxfWpEvKXuTFXLpcB2KvcleJaKjvZHbd+HYHW291KxGfxGkr4q6fiIuKXWUC3R2gsnHRwdvyUiYkLtHVvunP/6zNpSVMz2fbbvtX2vpJ0mn08erzvfIGy/T9ICSY+W9BhJZ7V5xsck26O2X2R7oaRTVMxjfoKkCyVdXGe2NmlrIep0m+23atWKpzepGA9ro3b/ObEGEdHWO0z04hBJcyPiAUmy/RFJ10r6QJ2hKvALFYtCToiIKzqOf9V22+9CMTQ5DBFsqmImwT4qCtT3JL2tjbs02b5f0i0qOtltyucrtf3CSY7KVWoHRsSy8vVGks7PYDet9SPiz3XnaLvWF9ic2N5W0maS7ljt1JaS7mbcq3lsf13S7pK+o+IX/L6SfizpTqm9c3xtL5B0dMcvjo0lnRgRr6s1WMvkMESwUgYXUE6WdMzqV6DLmREnq/vGIqjXwvIx6fs15ajaTpPFVVq5QrL10+uGLasCq/bvnbpZt/muEXGD7a1qyINpRMSCcmn2duWhm8sbBLbdiO2NI+KP0sqd3nKrF8nl9j/YRXUHGNBGU5ybPawQ6J3t56iYRXC7il/wW9o+PCIurzFWFU6UdKXtr6j4vl4m6YP1RmqfrMZgbT9G0u+jpd+U7XMkLYqIz612/PWS9o2IV9aTDGtie7GkV0fEzeXr7SSdExG71ZtscOV9uCYvHl8WETfWHKl1WtvBliu4PiLpD5KOl/QFFfMQR2wfFhHfrjNfn94maaHtQ/Tgu6/OknRgXaEwpZmTxVUqdg6zPbPOQBWaqVXDbrl8T0PV2g7W9k9V7O7zSEmflfT8iLiq3JfgnDavd7e9t1bdfXVpRCyqMw/WzPZZKjam/mJ56BBJo22/2m77aEn/pGI/Aqv4Bf/ZiDi11mAt0+YCe21E7Fw+vykintxxLosNRdB85Ybbb1axVFaSfiDpU23f+d/29Spu0/2X8vV6kq5kLvbaae0QgR685+tfVzvXzt8aaJVyH+LrylvenDTd17eMVXTmk8bV/lk6Q9fmAju3XMduFbuvT65pt6R164uFh4uIGLd9s+3HR8Sv6s5TsbMkXV3uRSBJL5F0Rn1x2qm1QwRAE9i+XNIuKlZv/WXyeES8qLZQFbG9m1ZtOvSDiLimzjxtRIEFBlBu5fcQEfHfw85StXIIZDM9+GaOuXXqSbV5iABoghdExLs6D9j+qKRWF1jbb1Fxa5/fatX4a0jiItdaoIMFBrCG21tf3/ar7bZvUXHH5t/XnaXN6GCBPtj+ZxV7D29TTmmatIGkK7q/q1XuUHE7cgyADhbog+1Hqrjj74clvbvj1H0R8Yd6UlXH9hmStlexv8fKOb0Rkdt0tKToYIE+RMSfJP3J9rtWO7V+uVl12y8G/ap8zCof6AMdLDCAjtt3T86/3lrFloVPrTVYRWyvL0nc3aA/dLDAACJix87X5R1m31RTnMrY3kHFBkqPKl//n6TDImJprcFahg4WqJjtG1YvvG1j+wpJx0bEZeXr50j6UEQ8o85cbUMHCwzA9ts7Xo5I2lXSXTXFqdJ6k8VVkiLi++WGL1gLFFhgMJ23JB9TcdX9azVlqdJttv9dxTCBJB0q6bYa87QSQwRABWw/IiLurztHVcq7yB6nB2/DOH/yHl3oDQUWGIDtPVTsMrV+RDze9lxJb4iI1l/owuAosMAAbF+t4oaAF0xu8m57SUTsMPU7m8n2BVOdz2GXsGFiDBYYUETcYT9oL+rxNX1tC+yhYpnsOZKuFptsD4QCCwzmDtvPkBTlzQ6PlnRTzZkGMUfSvpJeJenVKi7ancP81/4wRAAMoLxV/MclPU9Ft3eppKNz2IWqvN/YqySdIOm4iPhkzZFahwIL4EHKwvoPKorrVpIukHRmRPy6zlxtRIEF+mD7fVOcjog4fmhhKmT7bBW3jL9Y0rkRsaTmSK1GgQX6YPsdXQ6vJ+lISY+OiPWHHKkStie06t5incXBKn5xbDj8VO1FgQUGZHsDFRe3jpR0nqQTI+KeelOhCZhFAPTJ9qMkvV3SIZIWSNqVlU7oRIEF+mD7BEkHSfqspB3ZLxXdMEQA9KEcq/ybig1eGKtEVxRYAEhkpO4AAJArCiwAJEKBBYBEKLAAkMj/A0O1tYUt/qtTAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<Figure size 432x288 with 2 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "fig, ax = plt.subplots(1, 1)\n",
    "_ = co.heatmap(\n",
    "    df=-np.sign(cluster_overlap_stats.log_odds_ratio) * np.log10(q_values),\n",
    "    ax=ax,\n",
    "    xticklabels=True,\n",
    "    yticklabels=True,\n",
    "    yticklabel_side='left',\n",
    "    cmap=\"RdBu_r\",\n",
    "    show_guide=True,\n",
    "    norm=co.plotting.MidpointNormalize(vmin=-2, midpoint=0, vmax=3),\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "pharmaceutical-ideal",
   "metadata": {},
   "source": [
    "We have winners - this toy example clearly provided DMRs in proximity (or covering) genes from B cell, HSC and Neutrophil marker genesets. We can distinguish them from CFU-E and Monocyte genesets - but we can also see that there is some overlap between the CFU-E and HSC, and the Neutrophil and Monocyte annotations. Next we might want to check whether that is due to overlap between the genesets, or due to the presence of different genesets within our DMR clusters - but that's beyond the basic usage we wanted to introduce here :)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:regionset_profiler]",
   "language": "python",
   "name": "conda-env-regionset_profiler-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.8"
  },
  "toc-autonumbering": true
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
